components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/builders/ResponseBuilder.java

```diff
@@ -17,12 +17,44 @@
  */
 package org.wso2.carbon.identity.sso.saml.builders;
 
+import org.joda.time.DateTime;
 import org.opensaml.saml2.core.Response;
 import org.wso2.carbon.identity.base.IdentityException;
 import org.wso2.carbon.identity.sso.saml.dto.SAMLSSOAuthnReqDTO;
 
+import java.security.NoSuchAlgorithmException;
+
 public interface ResponseBuilder {
 
-    Response buildResponse(SAMLSSOAuthnReqDTO authnReqDTO, String sessionIndexId) throws IdentityException;
+    /**
+     * Build response using SAMLSSOAuthnReqDTO and session index.
+     *
+     * @param authnReqDTO    SAML sso authentication request DTO.
+     * @param sessionIndexId Session index ID.
+     * @return Built response object.
+     * @throws IdentityException
+     * @throws NoSuchAlgorithmException
+     * @deprecated Use {@link #buildResponse(SAMLSSOAuthnReqDTO, String, DateTime, String)} instead.
+     */
+    @Deprecated
+    Response buildResponse(SAMLSSOAuthnReqDTO authnReqDTO, String sessionIndexId)
+            throws IdentityException, NoSuchAlgorithmException;
+
+    /**
+     * Build response using SAMLSSOAuthnReqDTO, session index, initiated time and assertion ID.
+     *
+     * @param authnReqDTO    SAML sso authentication request DTO.
+     * @param sessionIndexId Session index ID.
+     * @param initTime       Initiated timestamp of the response.
+     * @param assetionId     SAML Assertion ID of the response.
+     * @return Built response object.
+     * @throws IdentityException
+     * @throws NoSuchAlgorithmException
+     */
+    default Response buildResponse(SAMLSSOAuthnReqDTO authnReqDTO, String sessionIndexId, DateTime initTime,
+                                   String assetionId) throws IdentityException, NoSuchAlgorithmException {
```

Comment:
Why 'NoSuchAlgorithm' needs to be thrown.
Shouldn't it be handled within the implementation and an IdentityException should be thrown if necessary to delegate the error to upper layer.

Reply:
Fixed.